Suggest using SELinux to harden the container host

jan-cerny opened the pull request:
Docker supports SELinux to separate containers from each other and from the host system.
Therefore I think the "Docker SSG profile" should require:
* enabling SELinux on the host in enforcing mode
* enabling SELinux support within Docker

This pull request:
* select some of the existing SElinux-related rules into the profile
* adds one new rule that checks whether Docker is configured to run with SELinux enabled For explanation of this rule, please see https://access.redhat.com/solutions/2429071

iankko commented:
@jan-cerny 
In overall looks good to me. Couple of nitpicks though:
* The OVAL definition for ```docker_selinux_enabled``` rule should depend on / extend ```package_docker_installed``` and ```service_docker_enabled``` rules. Because on default RHEL-7.2 clean installation, the ```docker``` rpm isn't installed. And in current state the rule returns "fail". If docker isn't installed or enabled, the configuration is safe,
* For the XCCDF the rule is currently placed in sebool group. Please move it outside of that
 ```
<Group id="selinux-booleans">
```
group. Maybe put the XCCDF rule right after the existing ```selinux_all_devicefiles_labeled```. Right now it looks in Workbench like docker selinux check would be one of those seboolean ones,
* And for the XCCDF text I would suggest split the description / rationale by 80 characters. Not all users might have such wide screen, the proposal has been created from. And it might be hard to read on those monitors.

Otherwise I don't have objections to the proposal. In overall looks good.

IMHO it can be merged once the above issues are fixed.

redhatrises commented on RHEL/7/input/xccdf/system/selinux.xml:
```diff
@@ -5405,4 +5405,22 @@ If this setting is enabled, it should be disabled.
+To enable the SELinux for Docker, the Docker service must be configured
+to run the Docker daemon with <tt>--selinux-enabled</tt> option.
+In <tt>/etc/sysconfig/docker</tt>  add or correct the following line to enable SELinux support in
+Docker daemon.
```
This should say the following:

To enable SELinux for the Docker service, the Docker service must be configured
to run the Docker daemon with the <tt>--selinux-enabled</tt> option.
In the <tt>/etc/sysconfig/docker</tt> configuration file, add or correct the following line to enable SELinux support in the Docker daemon:

redhatrises commented on RHEL/7/input/xccdf/system/selinux.xml:
```diff
@@ -5405,4 +5405,22 @@ If this setting is enabled, it should be disabled.
+Basically it uses a confined SELinux type <tt>svirt_lxc_net_t</tt> to isolate the container processes from the host, and it generates a unique Multi-Category Security label to allow SELinux to prevent one container process from attacking other container processes and content.
+<br /><br />
+If <tt>--selinux-enabled</tt> is not specified, docker does not execute SELinux code to set labels. This means the container processes will either run as <tt>docker_t</tt> or <tt>spc_t</tt> (depending on the version of policy you have installed). Both of these types are unconfined. SELinux will provide no security separation for these container processes.
+</rationale>
```
A better rationale IMO would be:

```If SELinux is not enabled, Docker does not use SELinux which means Docker runs unconfined, and SELinux will not provide security separation for Docker container processes. However enabling SELinux for the Docker service, prevents an attacker or rogue container from attacking other container processes and content as well as prevents taking over the host operating system. ```

iankko commented:
@jan-cerny 
> The OVAL definition for docker_selinux_enabled rule should depend on / extend package_docker_installed and service_docker_enabled rules. Because on default RHEL-7.2 clean installation, the docker rpm isn't installed. And in current state the rule returns "fail". If docker isn't installed or enabled, the configuration is safe,

To be more explicit WRT to this one, ```service_docker_enabled``` OVAL check already contains checks if docker package is installed. Thus I suggest to enhance your existing test like:

```
  <criteria operator="OR">
    <criteria operator="AND">
      <extend_definition comment='docker service active on system' 
      definition_ref="service_docker_enabled" />
      <criterion comment="Docker is configured to run with SELinux support" 
      test_ref="test_docker_selinux_enabled"/>
    </criteria>
    <extend_definition comment='docker service inactive on system'
    definition_ref="service_docker_enabled" negate="true" />
  </criteria>
```

And since ```docker_selinux_enabled``` rule will depend on ```service_docker_enabled``` rule, the corresponding OVAL should be moved to ```RHEL/7/input/oval/oval_5.11``` location to get included into benchmark only in moment, OVAL-5.11 (systemd*) is supported by oscap present on the system.

jan-cerny commented:
@iankko @redhatrises Hi, thank you for your your advice. It all sounds very reasonable.

I have amended my commits, specifically:
- moved the OVAL definition into the oval_5.11 directory
- extended the definition of `docker_selinux_enabled` rule to depend on `service_docker_enabled` rule
- moved the XCCDF rule before the SElinux Booleans group
- slightly edited the description of the XCCDF rule
- used the rationale suggested by @redhatrises 
- wrapped the description and the rationale by 80 characters

iankko commented on RHEL/7/input/xccdf/system/selinux.xml:
```diff
@@ -198,6 +198,26 @@ cannot properly restrict access to the device file.
+<pre>OPTIONS='--selinux-enabled'</pre>
+</description>
+<rationale>
+If SELinux is not explicitely enabled in the Docker deamon configuration,
```
s/deamon/daemon/

iankko commented:
> s/deamon/daemon/

Fixed the typo and merged manually (since this version looks good to me).
